agent/php_error.c

```diff
@@ -29,6 +31,98 @@ typedef struct _nr_php_exception_filter_t {
 
+/* Transient macro to free memory in nr_php_error_call_error_group_callback */
+#define FREE_MEM                    \
+  nr_php_zval_free(&txn_arr);       \
+  nr_php_zval_free(&error_arr);     \
+  nr_php_zval_free(&group_name_zv); \
+  nro_delete(agent_attributes);     \
+  nr_free(request_uri);             \
+  nr_free(path);                    \
+  nr_free(method);
+
+/*
+ * Purpose      Execute a user-defined PHP callback function that assigns a
+ *              custom group name to an error.
+ *
+ * @param       txn     The transaction object
+ * @param       klass   The error class string
+ * @param       message The error message string
+ * @param       file    The error file string
+ * @param       stack_json  The JSON stack trace string
+ */
+static void nr_php_error_call_error_group_callback(nrtxn_t* txn,
+                                                   char* klass,
+                                                   char* message,
+                                                   char* file,
+                                                   char* stack_json) {
+  zval* txn_arr = NULL;
+  zval* error_arr = NULL;
+  zval* group_name_zv = NULL;
+  zend_fcall_info fci;
+  zend_fcall_info_cache fcc;
+  nrobj_t* agent_attributes;
+  char* group_name_str = NULL;
+  char* request_uri = NULL;
+  char* path = NULL;
+  char* method = NULL;
+  int status_code = 0;
+
+  if (!is_error_callback_set()) {
+    return;
+  }
+
+  agent_attributes = nr_attributes_agent_to_obj(txn->attributes,
+                                                NR_ATTRIBUTE_DESTINATION_ALL);
+
+  request_uri = nr_strdup(nr_php_get_server_global("REQUEST_URI" TSRMLS_CC));
+  path = nr_strdup(txn->path);
+  method = nr_strdup(
+      nro_get_hash_string(agent_attributes, "request.method", NULL));
+  status_code = nr_php_http_response_code();
+  status_code = (status_code < 0) ? 0 : status_code;
+
+  txn_arr = nr_php_zval_alloc();
+  error_arr = nr_php_zval_alloc();
+
+  array_init(txn_arr);
+  array_init(error_arr);
+
+  nr_php_add_assoc_string(txn_arr, "request_uri", request_uri);
+  nr_php_add_assoc_string(txn_arr, "path", path);
+  nr_php_add_assoc_string(txn_arr, "method", method);
+  add_assoc_long(txn_arr, "status_code", (zend_long)status_code);
+
+  nr_php_add_assoc_string(error_arr, "klass", klass);
+  nr_php_add_assoc_string(error_arr, "message", message);
+  nr_php_add_assoc_string(error_arr, "file", file);
+  nr_php_add_assoc_string(error_arr, "stack", stack_json);
+
+  fci = NRPRG(error_group_user_callback).fci;
+  fcc = NRPRG(error_group_user_callback).fcc;
+
+  group_name_zv = nr_php_call_fcall_info(fci, fcc, txn_arr, error_arr);
+
+  if (!nr_php_is_zval_non_empty_string(group_name_zv)) {
+    nrl_debug(NRL_MISC,
+              "Error Group callback: Invalid return value (Non-Empty "
+              "String Required).");
+    FREE_MEM;
+    return;
+  }
+
+  group_name_str
+      = nr_strndup(Z_STRVAL_P(group_name_zv), NR_ATTRIBUTE_VALUE_LENGTH_LIMIT);
```

Comment:
Why NR_ATTRIBUTE_VALUE_LENGTH_LIMIT vs Z_STRLEN_P(group_name_zv) ?
The concern here is that if group_name_zv is less than NR_ATTRIBUTE_VALUE_LENGTH_LIMIT, we could be copying junk over.

Reply:
Looking at the implementation of `nr_strndup`, it closely follows the c library implementation of `strndup`, where *at most* n bytes are copied. So there shouldn't be any junk data at risk of being copied over- it searches for the null terminator to the string and only copied over the number of characters up to that null terminator.

Reply:
The rationale for using `NR_ATTRIBUTE_VALUE_LENGTH_LIMIT` is because any string longer than 255 characters must be truncated.

Reply:
zval string handling is special which is why they have the length attribute as well. They aren't guaranteed to be null terminated.
https://www.oreilly.com/library/view/programming-php/1565926102/ch14s06.html
```
Strings do not need to be null-terminated, but since most third-party libraries expect null-terminated strings it is a good idea to always null-terminate any string you create.
```
Could copy over the lesser of NR_ATTRIBUTE_VALUE_LENGTH_LIMIT or Z_STRLEN_P(group_name_zv).
something like 
```
nr_strndup(Z_STRVAL_P(group_name_zv), NR_ATTRIBUTE_VALUE_LENGTH_LIMIT < Z_STRLEN_P(group_name_zv) ? NR_ATTRIBUTE_VALUE_LENGTH_LIMIT : Z_STRLEN_P(group_name_zv));
```

Reply:
Ah the joys of PHP - added in 01c0f7f8fb2c9b0bf85fc156677c714fd3ea82d9